src/mbed_tools/build/_internal/config/assemble_build_config.py

```diff
@@ -3,88 +3,113 @@
+    previous_filter_data = None
+    app_data = None
     if mbed_app_file:
-        app_data = decode_json_file(mbed_app_file)
-        requires = app_data["requires"] if "requires" in app_data else []
-
-    while previous_cumulative_data != current_cumulative_data:
-        current_labels = current_cumulative_data.labels | current_cumulative_data.extra_labels
-        filtered_files = _filter_files(
-            mbed_lib_files,
-            current_labels,
-            current_cumulative_data.features,
-            current_cumulative_data.components,
-            requires,
+        # We need to obtain the file filter data from mbed_app.json so we can select the correct set of mbed_lib.json
+        # files to include in the config. We don't want to update the config object with all of the app settings yet
+        # as we won't be able to apply overrides correctly until all relevant mbed_lib.json files have been parsed.
+        app_data = source.from_file(
+            mbed_app_file, default_name="app", target_filters=FileFilterData.from_config(config).labels
         )
+        _get_app_filter_labels(app_data, config)
+
+    current_filter_data = FileFilterData.from_config(config)
+    while previous_filter_data != current_filter_data:
+        filtered_files = _filter_files(mbed_lib_files, current_filter_data)
+        for config_file in filtered_files:
+            config.update(source.from_file(config_file, target_filters=current_filter_data.labels))
+            # Remove any mbed_lib files we've already visited from the list so we don't parse them multiple times.
+            mbed_lib_files.remove(config_file)
+
+        previous_filter_data = current_filter_data
+        current_filter_data = FileFilterData.from_config(config)
+
+    # Apply mbed_app.json data last so config parameters are overriden in the correct order.
+    if app_data:
+        config.update(app_data)
+
+    return config
+
+
+def _get_app_filter_labels(mbed_app_data: dict, config: Config) -> None:
+    requires = mbed_app_data.get("requires")
+    if requires:
+        config["requires"] = requires
 
-        mbed_lib_sources = [Source.from_mbed_lib(file, current_labels) for file in filtered_files]
-        all_sources = [target_source] + mbed_lib_sources
-        if mbed_app_file:
-            all_sources = all_sources + [Source.from_mbed_app(mbed_app_file, current_labels)]
+    config.update(_get_file_filter_overrides(mbed_app_data))
 
-        previous_cumulative_data = current_cumulative_data
-        current_cumulative_data = CumulativeData.from_sources(all_sources)
 
-    _update_target_attributes(target_attributes, current_cumulative_data)
+def _get_file_filter_overrides(mbed_app_data: dict) -> dict:
+    return {
+        "overrides": [
+            override
+            for override in mbed_app_data.get("overrides", [])
+            if override.modifier or override.name == "requires"
```

Comment:
The `requires` is a standalone config option. Is there a scenario in which `requires` will be part of `overrides` config option?

Reply:
There's also `target.requires` which is part of the overrides.